rwkv.cpp

```diff
@@ -459,35 +460,45 @@ struct ggml_tensor * rwkv_layer_norm(ggml_context * ctx, struct ggml_tensor * x,
 
 // --- Implementation ---
 
-struct rwkv_layer_state {
-    struct ggml_tensor * ffn_xx;
-    struct ggml_tensor * att_xx;
-    struct ggml_tensor * att_aa;
-    struct ggml_tensor * att_bb;
-    struct ggml_tensor * att_pp;
-};
-
-struct rwkv_graph {
-    struct ggml_tensor * input_state;
-    std::unique_ptr<struct rwkv_layer_state []> input_layers;
-    std::unique_ptr<struct rwkv_layer_state []> output_layers;
-    struct ggml_tensor * token_index;
-    struct ggml_tensor * logits;
-    std::unique_ptr<struct ggml_cgraph> cgraph;
+struct rwkv_ctx_size {
+    size_t objects_count = 0;
+    size_t objects_size = 0;
+    size_t scratch_size = 0;
 };
 
-struct rwkv_ggml_guard {
+struct rwkv_ggml_context {
+    std::unique_ptr<uint8_t []> scratch;
     struct ggml_context * ctx;
-    ~rwkv_ggml_guard() { if (ctx) { ggml_free(ctx); } }
+
+    rwkv_ggml_context() {}
+
+    rwkv_ggml_context(struct rwkv_ctx_size size) {
+        scratch.reset(new(std::nothrow) uint8_t [size.scratch_size]);
+        if (!scratch) return;
```

Comment:
```suggestion
        if (!scratch) {
            return;
```
And places below :)

Reply:
sorry, keep forgetting that you do not care about vertical space usage of very simple if statements x3